### PR TITLE
[LT-707] Read test cases from db

### PR DESCRIPTION
**Pull request by hassantanveer97**
- If AssessmentQuestionXblock mapping has been established, test cases are read from DB
- Otherwise it falls back to reading test cases from the files
- Tested:
  - Server shell
  - Epic box
  - Code read from files
  - Code read from stdin

### CODE REVIEW

File: openassessment/xblock/job_sample_grader/code_grader.py [hunk 1]
```diff
@@ -152,47 +154,89 @@ def run_code(self, run_type, executor_id, source_code, problem_name):
...
-                        # This will allow us to use the same names for epicbox and server_shell.
-                        'name': input_file,
-                        'content': input_content,
-                    },
-                    'expected_output_file': {'name': expected_output_file},
-                }
+            try:
+                test_cases = json.loads(question.metadata)[run_type]
+            except:
+                test_cases = question.metadata[run_type]
```
Comment (HussainTaj-arbisoft) [inline openassessment/xblock/job_sample_grader/code_grader.py:170]:
```suggestion
            try:
                test_cases = json.loads(question.metadata).get(run_type, {})
            except:
                test_cases = question.metadata.get(run_type, {})
```

File: openassessment/xblock/job_sample_grader/code_grader.py [hunk 2]
```diff
@@ -152,47 +154,89 @@ def run_code(self, run_type, executor_id, source_code, problem_name):
...
+
+            for case_number in test_cases:
```
Comment (HussainTaj-arbisoft) [inline openassessment/xblock/job_sample_grader/code_grader.py:172]:
```suggestion
            for case_number in sorted(test_cases.keys()):
```

File: openassessment/xblock/job_sample_grader/code_grader.py [hunk 3]
```diff
@@ -239,6 +283,14 @@ def run_code(self, run_type, executor_id, source_code, problem_name):
...
                 }
 
+                input_file_name = case_file['input_file']['name']
+                output_file_name = case_file['expected_output_file']['name']
+
+                if question_mapping and os.path.isfile(input_file_name):
+                    os.remove(input_file_name)
+                if question_mapping and os.path.isfile(output_file_name):
+                    os.remove(output_file_name)
+
```
Comment (HussainTaj-arbisoft) [inline openassessment/xblock/job_sample_grader/code_grader.py:293]: Won't this delete the grader data in the case when a block mapping does not exist?

File: openassessment/xblock/job_sample_grader/code_grader.py [hunk 4]
```diff
@@ -152,47 +154,89 @@ def run_code(self, run_type, executor_id, source_code, problem_name):
...
+                    {
+                        'case_number': case_number,
+                        'input_file': {
+                            'name': input_file.name,
+                            'content': bytes(case['input'], 'utf-8'),
+                        },
+                        'expected_output_file': {'name': expected_output_file.name},
+                    }
+                )
+        else:
```
Comment (HussainTaj-arbisoft) [inline openassessment/xblock/job_sample_grader/code_grader.py:200]: Let's refactor this code. This function is huge.

Can we do something like self.get_test_case_files() ?

**Review by HussainTaj-arbisoft — CHANGES_REQUESTED** [review]

**Review by HussainTaj-arbisoft — COMMENTED** [review]
There's an empty `openassessment/xblock/job_sample_grader/tmp_data/empty.in` file committed.

**Comment by hassantanveer97** [comment]
> There's an empty `openassessment/xblock/job_sample_grader/tmp_data/empty.in` file committed.

@HussainTaj-arbisoft Its deletion of that empty file, not addition

**Review by HussainTaj-arbisoft — APPROVED** [review]

**Pull request merged**